### PR TITLE
Validator record update

### PR DESCRIPTION
**Pull request by g-r-a-n-t**
Addresses #81

### CODE REVIEW

File: beacon_chain/types/src/deposit.rs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+use super::deposit_parameters::DepositParameters;
+use super::{Hash256};
+
+pub struct Deposit {
+    pub merkle_branch: Hash256,
+    pub merkle_tree_index: u64,
+    pub deposit_data: DepositData
+}
+
+pub struct DepositData {
```
Comment (g-r-a-n-t) [inline beacon_chain/types/src/deposit.rs]: @paulhauner Is creating a separate struct for deposit_data reasonable?

**Comment by paulhauner** [comment]
Hey @g-r-a-n-t, I just merged a PR that adds SSZ enc/decoding to `ValidatorRecord`. Apologies for the conflict.

**Comment by g-r-a-n-t** [comment]
No worries! @paulhauner

**Comment by g-r-a-n-t** [comment]
This should be ready for review now @paulhauner :)

File: beacon_chain/validator_induction/src/inductor.rs [hunk 2]
```diff
@@ -1,278 +1,187 @@
...
-        None
+pub fn process_deposit(
+    state: &mut BeaconState,
+    deposit: &Deposit,
+    spec: &ChainSpec) 
+-> Result<usize, ValidatorInductionError> {
+    let deposit_input = &deposit.deposit_data.deposit_input;
+    let deposit_data = &deposit.deposit_data;
+        
+    if !verify_proof_of_possession(&deposit_input.proof_of_possession, &deposit_input.pubkey) {
```
Comment (g-r-a-n-t) [inline beacon_chain/validator_induction/src/inductor.rs]: I wasn't able to figure out how this should look based on the spec. If it's not a big deal, it might be best to update this in another PR.

Comment (paulhauner) [reply beacon_chain/validator_induction/src/inductor.rs]: Sorry, which part are you unsure about?

Comment (g-r-a-n-t) [reply beacon_chain/validator_induction/src/inductor.rs]: I wasn't sure how implement the `validate_proof_of_possession` helper method as defined in the [spec](https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#routine-for-processing-deposits), partly because there is no `domain` parameter in our BLS signature, and also because there doesn't seem to be an implementation of `hash_tree_root`. Apologies for the lack of clarity. It looks to me like this part of the deposit function should get refactored at a later point.

Comment (paulhauner) [reply beacon_chain/validator_induction/src/inductor.rs]: Oh yep, I'm with you now.

I agree it'll need a refactor. I think it's fine to leave it like this as long as we drop a `TODO` in there referencing #91 and #70 :)

**Comment by paulhauner** [comment]
Hey mate, this is first off the rank for me on Monday :)

File: beacon_chain/types/src/deposit.rs [hunk 3]
```diff
@@ -10,6 +11,29 @@ pub struct Deposit {
     pub deposit_data: DepositData,
 }
 
+impl Deposit {
+    pub fn zero_with_rand_keypair() -> Self{
```
Comment (paulhauner) [inline beacon_chain/types/src/deposit.rs]: Sorry I hadn't communicated this, but I'd like to remove the `zero_with_rand_keypair()` functions in favor of the `TestRandom` trait.

I had a quick look and it should be pretty easy to swap out.

Comment (g-r-a-n-t) [reply beacon_chain/types/src/deposit.rs]: That's much cleaner.

File: beacon_chain/genesis/src/beacon_state.rs [hunk 4]
```diff
@@ -11,26 +10,13 @@ pub enum Error {
...
-        for registration in &spec.initial_validators {
-            let _ = inductor.induct(&registration, ValidatorStatus::Active);
-        }
-        inductor.to_vec()
-    };
-
     /*
      * Assign the validators to shards, using all zeros as the seed.
      *
      * Crystallizedstate stores two cycles, so we simply repeat the same assignment twice.
```
Comment (ralexstokes) [inline beacon_chain/genesis/src/beacon_state.rs]: we should remove this ref to `crystallized_state` -- i can make another PR if thats easier, i think the PR that did this clean up was already merged

Comment (g-r-a-n-t) [reply beacon_chain/genesis/src/beacon_state.rs]: I removed this line, if that's what you were hoping for.

File: beacon_chain/types/src/deposit.rs (hunk 3 above)
Comment (g-r-a-n-t) [reply beacon_chain/types/src/deposit.rs]: I'll remove the one in `ValidatorRecord` as well

**Comment by paulhauner** [comment]
The spec has changed again! I think we should just merge this in and do another PR once we think the struct has stabilized a bit! Thoughts @ralexstokes?

Australian's have an saying of "painting the harbor bridge" (it's a large bridge). By the time you've finished the last section, the first section needs painting again :laughing:

**Comment by ralexstokes** [comment]
looks awesome @g-r-a-n-t thanks for making those updates :)

oh yes -- lets go ahead and merge @paulhauner  -- we will never merge if we wait for the spec ;)
